### DISCUSSION TITLE
Some questions about external_petsc_problem1.0?

### DISCUSSION THREAD

**Discussion by Wang-Yihu** [idaholab/moose#29368] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Q1: What does the problem the `\modules\external_petsc_solver\test\tests\external_petsc_problem\petsc_transient_as_sub.i `describe? 
Q2: What makes the Picard iteration? What's the meaning of the Picard residual in the `moose_as_parent.i` and `petsc_transient_as_sub.i`?

Recently, I still study the code in `\modules\external_petsc_solver\`. I encounter some puzzles...
1.
It seems that the `moose_as_parent.i` tells a story of 

∂u/∂t - ∇·∇u – 10000v = 0

And v comes from the transferred value from the u in `petsc_transient_as_sub.i`. But what does the `petsc_transient_as_sub.i` say? From the code comment of `PETScDiffusionFDM.C`, it seems that the govern function is

∂u/∂t = ∂2u/∂x2 + ∂2u/∂y2 

However, I have two questions from its other comments. 
1.1 the comment says  At t=0: u(x,y) = exp(c*r*r*r), if r=PetscSqrtReal((x-.5)*(x-.5) + (y-.5)*(y-.5)) < .125. If that is right, at t = 0, x = 0.5, y = 0.5, the u must be 1, right? However, the initial condition of u at the center point from the *.e file(moose_as_parent_out_sub_app0.e) generated by MOOSE is 0 not 1... What's the wrong with it?
1.2 the comment also says that
> Boundary conditions:
>    Drichlet BC:
>    At x=0, x=1, y=0, y=1: u = 0.0
> 
>    Neumann BC:
>    At x=0, x=1: du(x,y,t)/dx = 0
>    At y=0, y=1: du(x,y,t)/dy = 0

I also find similar things in the code `PETScDiffusionFDM.C`
```C++
 for (j = ys; j < ys + ym; j++)
  {
    for (i = xs; i < xs + xm; i++)
    {
      nc = 0;
      row.j = j;
      row.i = i;
      if (PETSC_TRUE && (i == 0 || i == Mx - 1 || j == 0 || j == My - 1))
      {
        col[nc].j = j;
        col[nc].i = i;
        vals[nc++] = 1.0;
      }
      else if (PETSC_FALSE && i == 0)
      { /* Left Neumann */
        col[nc].j = j;
        col[nc].i = i;
        vals[nc++] = 1.0;
        col[nc].j = j;
        col[nc].i = i + 1;
        vals[nc++] = -1.0;
      }
      else if (PETSC_FALSE && i == Mx - 1)
      { /* Right Neumann */
        col[nc].j = j;
        col[nc].i = i;
        vals[nc++] = 1.0;
        col[nc].j = j;
        col[nc].i = i - 1;
        vals[nc++] = -1.0;
      }
      else if (PETSC_FALSE && j == 0)
      { /* Bottom Neumann */
        col[nc].j = j;
        col[nc].i = i;
        vals[nc++] = 1.0;
        col[nc].j = j + 1;
        col[nc].i = i;
        vals[nc++] = -1.0;
      }
      else if (PETSC_FALSE && j == My - 1)
      { /* Top Neumann */
        col[nc].j = j;
        col[nc].i = i;
        vals[nc++] = 1.0;
        col[nc].j = j - 1;
        col[nc].i = i;
        vals[nc++] = -1.0;
      }
      else
      { /* Interior */
        col[nc].j = j - 1;
        col[nc].i = i;
        vals[nc++] = -sy;
        col[nc].j = j;
        col[nc].i = i - 1;
        vals[nc++] = -sx;
        col[nc].j = j;
        col[nc].i = i;
        vals[nc++] = 2.0 * (sx + sy) + a;
        col[nc].j = j;
        col[nc].i = i + 1;
        vals[nc++] = -sx;
        col[nc].j = j + 1;
        col[nc].i = i;
        vals[nc++] = -sy;
      }
      ierr = MatSetValuesStencil(Jpre, 1, &row, nc, col, vals, INSERT_VALUES);
      CHKERRQ(ierr);
    }
  }
```
However, I find there is a `PETSC_FALSE` near these Neumaan comment, it seems that program statements of `else if (PETSC_FALSE && i == 0)`, `else if (PETSC_FALSE && i == Mx - 1)`, `else if (PETSC_FALSE && j == 0)`,` else if (PETSC_FALSE && j == My - 1)`, will not be executed by default... is this a paradox???
I will show Q2 in another discusssion...

**Comment by GiudGiud** [comment, marked as answer]
> Q1: What does the problem the \modules\external_petsc_solver\test\tests\external_petsc_problem\petsc_transient_as_sub.i describe?

You have to look into the ExternalPetscProblem class to see what it's doing. It's doing implicit Euler for time integration.
then there is a call to ` externalPETScDiffusionFDMSolve` which is defined here
https://github.com/idaholab/moose/blob/next/modules/external_petsc_solver/src/petscsolver/PETScDiffusionFDM.C

It's a finite difference diffusion problem. 

>  What makes the Picard iteration? What's the meaning of the Picard residual in the moose_as_parent.i and petsc_transient_as_sub.i?

See the other post. Picard iteration is from the FixedPointSolve created in the main app

> he comment says At t=0: u(x,y) = exp(crrr), if r=PetscSqrtReal((x-.5)(x-.5) + (y-.5)*(y-.5)) < .125. If that is right, at t = 0, x = 0.5, y = 0.5, the u must be 1, right? However, the initial condition of u at the center point from the *.e file(moose_as_parent_out_sub_app0.e) generated by MOOSE is 0 not 1... What's the wrong with it?

Dont read too much into the comments. The code might have evolved without updating the comments. This is one of the main flaws of commenting code.
